pkgs/tools/security/ghidra/build.nix

```diff
@@ -0,0 +1,151 @@
+{ stdenv
+, fetchzip
+, fetchurl
+, fetchFromGitHub
+, lib
+, gradle
+, perl
+, makeWrapper
+, openjdk11
+, unzip
+, makeDesktopItem
+, autoPatchelfHook
+, icoutils
+, xcbuild
+}:
+
+let
+  pkg_path = "$out/lib/ghidra";
+  pname = "ghidra";
+  version = "10.1.1";
+
+  src = fetchFromGitHub {
+    owner = "NationalSecurityAgency";
+    repo = "Ghidra";
+    rev = "Ghidra_${version}_build";
+    sha256 = "sha256-0hj9IVvTxgStCbfnTzqeKD+Q5GnGowDsIkMvk2GqJqY=";
+  };
+
+  desktopItem = makeDesktopItem {
+    name = "ghidra";
+    exec = "ghidra";
+    icon = "ghidra";
+    desktopName = "Ghidra";
+    genericName = "Ghidra Software Reverse Engineering Suite";
+    categories = "Development;";
+  };
+
+  # fake build to pre-download deps into fixed-output derivation
+  # Taken from mindustry derivation.
+  deps = stdenv.mkDerivation {
+    pname = "${pname}-deps";
+    inherit version src;
+
+    nativeBuildInputs = [ gradle perl ];
+    buildPhase = ''
+    cat >>build.gradle <<HERE
+task resolveDependencies {
+  doLast {
+    project.rootProject.allprojects.each { subProject ->
+      subProject.buildscript.configurations.each { configuration ->
+        resolveConfiguration(subProject, configuration)
+      }
+      subProject.configurations.each { configuration ->
+        resolveConfiguration(subProject, configuration)
+      }
+    }
+  }
+}
+void resolveConfiguration(subProject, configuration) {
+  if (configuration.canBeResolved) {
+    logger.info("Resolving project {} buildscript config {}", subProject.name, configuration.name)
+    configuration.resolve()
+  }
+}
+HERE
+      export GRADLE_USER_HOME=$(mktemp -d)
+
+      # First, fetch the static dependencies.
+      gradle --no-daemon --info -Dorg.gradle.java.home=${openjdk11} -I gradle/support/fetchDependencies.gradle init
+
+      # Then, fetch the maven dependencies.
+      gradle --no-daemon --info -Dorg.gradle.java.home=${openjdk11} resolveDependencies
+    '';
+    # perl code mavenizes pathes (com.squareup.okio/okio/1.13.0/a9283170b7305c8d92d25aff02a6ab7e45d06cbe/okio-1.13.0.jar -> com/squareup/okio/okio/1.13.0/okio-1.13.0.jar)
+    installPhase = ''
+      find $GRADLE_USER_HOME/caches/modules-2 -type f -regex '.*\.\(jar\|pom\)' \
+        | perl -pe 's#(.*/([^/]+)/([^/]+)/([^/]+)/[0-9a-f]{30,40}/([^/\s]+))$# ($x = $2) =~ tr|\.|/|; "install -Dm444 $1 \$out/maven/$x/$3/$4/$5" #e' \
+        | sh
+      find $GRADLE_USER_HOME/caches/modules-2 -type f -regex '.*\.exe' \
+        | perl -pe 's#(.*/([^/]+)/([^/]+)/([^/]+)/[0-9a-f]{30,40}/([^/\s]+))$# ($x = $2) =~ tr|\.|/|; "install -Dm555 $1 \$out/maven/$x/$3/$4/$5" #e' \
+        | sh
+      cp -r dependencies $out/dependencies
+    '';
+    outputHashAlgo = "sha256";
+    outputHashMode = "recursive";
+    outputHash = "sha256-tiJpctga2ddPJbO9qvYQBpDmpEn6ncCjfDIxg8YWs5U=";
+  };
+
+  fixedDeps = stdenv.mkDerivation {
+    pname = "${pname}-fixeddeps";
+    inherit version;
+
+    nativeBuildInputs = [ autoPatchelfHook ];
```

Comment:
This needs to be omitted in the darwin case.

Reply:
Hmm. So this is a bit of a kludge (which I should really document). It's used to patch the `protoc` binary that's given to us by maven. Basically in the maven repo, there's a prebuilt protoc binary that is used during the build, and we use autoPatchelfHook to fix it up for nixos usage. Ideally we should use our own protoc binary, but I haven't found where in gradle to pass our own binary. I think it's only prebuilt for x86 and arm platforms.

As for why it's not necessary on macos, well macos has all the necessary native libraries, so it's not really necessary to fix it up.

Reply:
So, turns out, it's possible to set the protoc binary. https://github.com/google/protobuf-gradle-plugin#customizing-protobuf-compilation

So I could probably omit copying prebuilt binaries and just patch the build.gradle to use our protoc instead.

Reply:
Another reason it might "just work" on macos is I'm building with the sandbox disabled (as hydra does)

Reply:
Ghidra doesn't use the protobuf gradle plugin, but has its own custom task: https://github.com/NationalSecurityAgency/ghidra/blob/c51183f1a262c80d45aac48d4f3343c67b6bd41a/Ghidra/Debug/Debugger-gadp/build.gradle#L77-L96

I'll try to remove this task and change it to use the gradle plugin, and upstream that, as it'll make the packaging much easier: https://github.com/NationalSecurityAgency/ghidra/pull/3813

Reply:
While waiting for an answer on the ghidra MR, I included the changes in the form of a patchfile.